### PR TITLE
When you like a repost, put likes on the entire hierarchy of repost posts recursively

### LINKED ISSUE #159: Лайк на репост не ставит лайк на оригинальный пост
Ну всё уже понятно из названия, в VK ставя лайк на репост - ставится лайк и на оригинальный пост, в OVK такого нет, но стоит реализовать.

### PR DESCRIPTION
**Pull request by maksalees**
If you like a repost, all posts from the hierarchy will be liked. If you remove a like from a post, likes will be removed from all posts from the hierarchy. I don’t know what the load will be in the case of large hierarchies (for example, 20 posts), but with small hierarchies (5 posts) everything works fine. Fixes #159

### CODE REVIEW

**Review by celestora — CHANGES_REQUESTED** [review]
Would it be possible for you to implement a configurable recursion depth limit to prevent DoS-attacks?

**Comment by maksalees** [comment]
> Would it be possible for you to implement a configurable recursion depth limit to prevent DoS-attacks?

Yes. But... What limitation should be the default? After raising the limit, just turn off the repost button or should it only concern likes?

**Comment by veselcraft** [comment]
> Yes. But... What limitation should be the default? After raising the limit, just turn off the repost button or should it only concern likes?

Three is OK

**Comment by maksalees** [comment]
> > Yes. But... What limitation should be the default? After raising the limit, just turn off the repost button or should it only concern likes?
> 
> Three is OK

It seems to me that 5 or 10 is better

**Comment by celestora** [comment]
> After raising the limit, just turn off the repost button or should it only concern likes?

If recursion depth will exceed the limit, just return without recursing further :3

**Comment by celestora** [comment]
> > Yes. But... What limitation should be the default? After raising the limit, just turn off the repost button or should it only concern likes?
> 
> Three is OK

I was thinking more of 20, but ok

**Comment by maksalees** [comment]
> Would it be possible for you to implement a configurable recursion depth limit to prevent DoS-attacks?

@celestora Done

File: Web/Models/Entities/Post.php [hunk 1]
```diff
@@ -153,10 +158,11 @@ function setLike(bool $liked, User $user): void
...
 
-        foreach($this->getChildren() as $attachment) {
-            if($attachment instanceof Post)
-                $attachment->setLike($liked, $user);
-        }
+        if($depth < ovkGetQuirk("wall.repost-liking-recursion-limit"))
+            foreach($this->getChildren() as $attachment) {
+                if($attachment instanceof Post)
+                    $attachment->setLikeRecurse($liked, $user, $depth + 1);
+            }
```
Comment (celestora) [inline Web/Models/Entities/Post.php]: Would you kindly remove redundant "{" and "}"?

File: Web/Models/Entities/Post.php [hunk 2]
```diff
@@ -132,13 +132,18 @@ function toggleLike(User $user): bool
...
 
         return $liked;
     }
 
     function setLike(bool $liked, User $user): void
+    {
+        $this->setLikeRecurse($liked, $user, 1);
+    }
+
+    private function setLikeRecurse(bool $liked, User $user, int $depth): void
```
Comment (celestora) [inline Web/Models/Entities/Post.php]: By convention, private functions are on the top of the class body + methods should be defined before they're mentioned in others. Also, rename method to `likeRecursively` or something like this, would you kindly?

**Review by celestora — CHANGES_REQUESTED** [review]

File: Web/Models/Entities/Post.php (hunk 1 above)
Comment (maksalees) [reply Web/Models/Entities/Post.php]: Done

File: Web/Models/Entities/Post.php (hunk 2 above)
Comment (maksalees) [reply Web/Models/Entities/Post.php]: Done

**Comment by maksalees** [comment]
@celestora I made the changes you requested

**Comment by celestora** [comment]
> @celestora I made the changes you requested

sugoi nanodesu

**Comment by veselcraft** [comment]
@celestora naze anata wa nohingo o hanashimasu ka

**Pull request merged**